samanthas_telegram_bot/api_queries/smalltalk.py

```diff
@@ -0,0 +1,208 @@
+                f"{user_data.comment}\n- SmallTalk assessment not finished\nCheck {result.url}"
+            )
+            return None
+        elif result.status == SmalltalkTestStatus.RESULTS_NOT_READY:
+            if attempts > SMALLTALK_MAX_ATTEMPTS_TO_GET_RESULTS:
+                total_seconds_waiting = (
+                    SMALLTALK_MAX_ATTEMPTS_TO_GET_RESULTS
+                    * SMALLTALK_TIMEOUT_IN_SECS_BETWEEN_ATTEMPTS
+                )
+                await log_and_notify(
+                    bot=context.bot,
+                    logger=logger,
+                    level=LoggingLevel.ERROR,
+                    text=(
+                        f"Chat {user_data.chat_id}: SmallTalk results for {user_data.first_name} "
+                        f"{user_data.last_name} still not ready after "
+                        f"{total_seconds_waiting / 60} minutes. "
+                        f"Interview ID {user_data.student_smalltalk_test_id}."
+                    ),
+                )
+                user_data.comment = (
+                    f"{user_data.comment}\n- SmallTalk assessment results were not ready\n"
+                    f"Interview ID {user_data.student_smalltalk_test_id}"
+                )
+
+            logger.info(f"Chat {user_data.chat_id}: SmallTalk results not ready. Waiting...")
+            attempts += 1
+            await asyncio.sleep(SMALLTALK_TIMEOUT_IN_SECS_BETWEEN_ATTEMPTS)
+        else:
+            await log_and_notify(
+                bot=context.bot,
+                logger=logger,
+                level=LoggingLevel.INFO,
+                text=(
+                    f"Chat {user_data.chat_id}: Received [SmallTalk results for "
+                    f"{user_data.first_name} {user_data.last_name}]({result.url})"
+                ),
+                parse_mode_for_admin_group_message=ParseMode.MARKDOWN_V2,
+            )
+            return result
+
+
+async def get_json_with_results(test_id: str) -> bytes:
+    async with httpx.AsyncClient() as client:
+        response = await client.get(
+            url=SMALLTALK_URL_GET_RESULTS,
+            headers=HEADERS,
+            params={
+                "id": test_id,
+                "additional_fields": (
+                    "detailed_scores,strength_weaknesses,problem_statuses,problem_titles"
+                ),
+            },
+        )
+
+    logger.info(
+        f"Request headers: {response.request.headers}. "
+        f"Response: {response.status_code=}, {response.headers=}, {response.content=}"
+    )
+    return response.content
+
+
+def process_smalltalk_json(json_data: bytes) -> SmalltalkResult | None:
+    def level_is_undefined(str_: str) -> bool:
+        return str_.lower().strip() == "undefined"
+
+    try:
+        loaded_data = json.loads(json_data)
+    except json.decoder.JSONDecodeError:
+        logger.error(f"Could not load JSON from {json_data=}")
+        return None
+
+    status = loaded_data["status"]
```

Comment:
What if there are no "status" key in response? Or maybe smalltalk response was 500? 

You should never believe to external service.

Reply:
Fixed